src/pages/index.js

```diff
@@ -103,101 +113,90 @@ const renCard = (data) => {
+        if (isLiked === true) {
+          api
+            .dislikeCard(card)
+            .then(() => {})
+            .catch((err) => {
+              console.error(err);
+            });
+        } else {
+          api
+            .likeCard(card)
+            .then(() => {})
+            .catch((err) => {
+              console.error(err);
+            });
+        }
+      },
     },
     cardSelector
   );
 
   cardSection.addItem(cardElement.getView());
 };
-// render card function
-
-const renderCard = (data) => {
-  const cardElement = new Card(
-    {
-      data,
-      handleImagePreview: (imageData) => {
-        previewImage.open(imageData);
-      },
-      handleDeleteIcon: (cardElement) => {
-        const deleteCardPopup = new PopupDelete(
-          deletePopupSelector,
-          deleteButtonSelector,
 
-          () => {
-            // console.log(cardElement.id);
+// popup with form - edit avatar
 
-            api
-              .deleteCard(cardElement)
-              .then(() => {
-                if (cardElement) {
-                  cardElement.remove();
-                }
-              })
-              .catch((err) => {
-                console.error(err);
-              });
-          }
-        );
-        deleteCardPopup.open();
-      },
-    },
-    cardSelector
-  );
-  cardSection.addItem(cardElement.getView());
-};
+const editAvatarPopUp = new PopupWithForm(formAvatarSelector, (info) => {
+  editAvatarPopUp.uploadingInfo(true);
+  api
+    .editUserAvatar(info)
+    .then((info) => {
+      setAvatar(info);
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      editAvatarPopUp.uploadingInfo(false);
+      editAvatarPopUp.close();
+    });
+});
 
 // //popup with form - add card
 
-const addCardPopup = new PopupWithForm(formAddSelector, (data) => {
-  const name = data.name;
-  const link = data.link;
-  const id = data._id;
-  const cardData = {
-    name,
-    link,
-    id,
-  };
-
-  newCard(cardData);
-});
+const addCardPopup = new PopupWithForm(formAddSelector, newCard);
 
 function newCard(cardData) {
+  addCardPopup.uploadingInfo(true);
   api
     .addNewCard(cardData)
     .then((addedCardData) => {
       console.log(addedCardData);
 
-      renCard(addedCardData);
-
-      addCardPopup.close();
+      renderCard(addedCardData);
     })
 
     .catch((err) => {
       console.error(err);
+    })
+    .finally(() => {
+      addCardPopup.uploadingInfo(false);
+      addCardPopup.close();
     });
 }
 
-// function addCardSuccess(cardData) {
-//   renderCard(cardData);
-// }
-
 // // popup with form - edit profile info
 
 const editProfile = new PopupWithForm(formEditSelector, (info) => {
+  editProfile.uploadingInfo(true);
   api
     .editUserInfo(info)
     .then((info) => {
+      setProfileInfo(info);
+
       return info;
     })
     .catch((err) => {
       console.error(err);
+    })
+    .finally(() => {
+      editProfile.uploadingInfo(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```